# Análise de dados com Pandas e Seaborn

O Pandas é uma biblioteca Python de código aberto para análise de dados. Ela fornece ferramentas de análise de dados e estruturas de dados de alta performance e fáceis de usar. Já o Seaborn é uma biblioteca Python de visualização de dados baseada no matplotlib. Ela fornece uma interface de alto nível para desenhar gráficos estatísticos atraentes e informativos. Sua vantagem sobre o Matplotlib puro é que Seaborn fornece padrões de plotagem estatística que podem ser facilmente ativados com apenas alguns comandos.

A base de dados utilizada aqui é de churn de clientes de uma empresa de telecomunicações. O churn é uma métrica que indica a taxa de evasão de clientes. Neste caso, o churn é definido como a perda de clientes para a concorrência ou por desinteresse em continuar com o serviço. O objetivo é analisar os dados e identificar os principais fatores que levam um cliente a cancelar o serviço.

Temos o seguinte dicionário de dados:

- `id_cliente`: identificador único do cliente
- `genero`: gênero do cliente
- `idoso`: indica se o cliente é idoso (mais de 65 anos) ou não
- `tem_parceiro`: indica se o cliente tem parceiro ou não
- `tem_dependentes`: indica se o cliente tem dependentes ou não
- `meses_na_empresa`: quantidade de meses que o cliente permaneceu na empresa
- `servico_telefone`: indica se o cliente tem serviço de telefone ou não
- `multiplas_linhas`: indica se o cliente tem múltiplas linhas ou não
- `servico_internet`: indica se o cliente tem serviço de internet (DSL ou fibra óptica) ou não
- `servico_seguranca`: indica se o cliente tem serviço de segurança online ou não
- `servico_backup`: indica se o cliente tem serviço de backup online ou não
- `servico_protecao_equipamento`: indica se o cliente tem serviço de proteção do equipamento de conexão ou não
- `servico_suporte_tecnico`: indica se o cliente tem serviço de suporte técnico prioritário ou não
- `streaming_tv`: indica se o cliente tem serviço de streaming de TV ou não
- `streaming_filmes`: indica se o cliente tem serviço de streaming de filmes ou não
- `contrato`: tipo de contrato do cliente (mensal, anual ou dois anos)
- `fatura_digital`: indica se o cliente recebe a fatura por e-mail ou não
- `forma_pagamento`: forma de pagamento do cliente (cheque, boleto, transferência bancária ou cartão de crédito)
- `valor_mensal`: valor mensal do serviço contratado
- `total_gasto`: valor total gasto pelo cliente
- `churn`: indica se o cliente cancelou o serviço no mês anterior ou não



In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

ARQUIVO_DADOS = 'churn_clientes.csv'

df_churn = pd.read_csv(ARQUIVO_DADOS)
df_churn.head()


id_cliente     genero idoso tem_parceiro tem_dependentes  meses_na_empresa  \
0  7590-VHVEG   Feminino   Não          Sim             Não                 1   
1  5575-GNVDE  Masculino   Não          Não             Não                34   
2  3668-QPYBK  Masculino   Não          Não             Não                 2   
3  7795-CFOCW  Masculino   Não          Não             Não                45   
4  9237-HQITU   Feminino   Não          Não             Não                 2   

  servico_telefone multiplas_linhas servico_internet servico_seguranca  ...  \
0              Não              Não              DSL               Não  ...   
1              Sim              Não              DSL               Sim  ...   
2              Sim              Não              DSL               Sim  ...   
3              Não              Não              DSL               Sim  ...   
4              Sim              Não     Fibra optica               Não  ...   

  servico_protecao_equipamento servico_suporte_tecnico streaming_tv  \
0                          Não                     Não          Não   
1                          Sim                     Não          Não   
2                          Não                     Não          Não   
3                          Sim                     Sim          Não   
4                          Não                     Não          Não   

  streaming_filmes contrato fatura_digital forma_pagamento valor_mensal  \
0              Não   Mensal            Sim          Boleto        29.85   
1              Não    Anual            Não          Cheque        56.95   
2              Não   Mensal            Sim          Cheque        53.85   
3              Não    Anual            Não   Transferencia        42.30   
4              Não   Mensal            Sim          Boleto        70.70   

   total_gasto  churn  
0        29.85    Não  
1      1889.50    Não  
2       108.15    Sim  
3      1840.75    Não  
4       151.65    Sim  

[5 rows x 21 columns]

In [2]:
df_churn.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   id_cliente                    7043 non-null   object 
 1   genero                        7043 non-null   object 
 2   idoso                         7043 non-null   object 
 3   tem_parceiro                  7043 non-null   object 
 4   tem_dependentes               7043 non-null   object 
 5   meses_na_empresa              7043 non-null   int64  
 6   servico_telefone              7043 non-null   object 
 7   multiplas_linhas              7043 non-null   object 
 8   servico_internet              7043 non-null   object 
 9   servico_seguranca             7043 non-null   object 
 10  servico_backup                7043 non-null   object 
 11  servico_protecao_equipamento  7043 non-null   object 
 12  servico_suporte_tecnico       7043 non-null   object 
 13  str

In [3]:
df_churn.describe() 

meses_na_empresa  valor_mensal  total_gasto
count       7043.000000   7043.000000  7043.000000
mean          32.371149     64.761692  2281.916928
std           24.559481     30.090047  2265.270398
min            0.000000     18.250000    18.800000
25%            9.000000     35.500000   402.225000
50%           29.000000     70.350000  1397.475000
75%           55.000000     89.850000  3786.600000
max           72.000000    118.750000  8684.800000

In [4]:
df_churn.describe(exclude='number') 

id_cliente     genero idoso tem_parceiro tem_dependentes  \
count         7043       7043  7043         7043            7043   
unique        7043          2     2            2               2   
top     7590-VHVEG  Masculino   Não          Não             Não   
freq             1       3555  5901         3641            4933   

       servico_telefone multiplas_linhas servico_internet servico_seguranca  \
count              7043             7043             7043              7043   
unique                2                2                3                 2   
top                 Sim              Não     Fibra optica               Não   
freq               6361             4072             3096              5024   

       servico_backup servico_protecao_equipamento servico_suporte_tecnico  \
count            7043                         7043                    7043   
unique              2                            2                       2   
top               Não                          Não                     Não   
freq             4614                         4621                    4999   

       streaming_tv streaming_filmes contrato fatura_digital forma_pagamento  \
count          7043             7043     7043           7043            7043   
unique            2                2        3              2               4   
top             Não              Não   Mensal            Sim          Boleto   
freq           4336             4311     3875           4171            2365   

       churn  
count   7043  
unique     2  
top      Não  
freq    5174

Agora sabemos que a maioria dos clientes:

- é do sexo masculino
- é mais jovem que 65 anos
- não tem parceiro e nem dependentes
- tem serviço de telefone, com uma única linha
- tem serviço de internet por fibra óptica
- não tem serviços online (segurança, backup, proteção de dispositivo)
- não tem suporte técnico
- não tem serviços de streaming
- tem contratos mensais
- tem faturamento digital e pagam com boleto

A variável alvo *churn* tem "Não" como o valor mais frequente, ocorrendo 5174 vezes.


In [5]:
df_churn['churn'].value_counts()

Não    5174
Sim    1869
Name: churn, dtype: int64

In [6]:
df_churn['churn'].value_counts(normalize=True) * 100 

Não    73.463013
Sim    26.536987
Name: churn, dtype: float64

In [7]:
# drop da coluna id_cliente
df_churn = df_churn.drop(columns='id_cliente', axis=1)
df_churn.head() 

genero idoso tem_parceiro tem_dependentes  meses_na_empresa  \
0   Feminino   Não          Sim             Não                 1   
1  Masculino   Não          Não             Não                34   
2  Masculino   Não          Não             Não                 2   
3  Masculino   Não          Não             Não                45   
4   Feminino   Não          Não             Não                 2   

  servico_telefone multiplas_linhas servico_internet servico_seguranca  \
0              Não              Não              DSL               Não   
1              Sim              Não              DSL               Sim   
2              Sim              Não              DSL               Sim   
3              Não              Não              DSL               Sim   
4              Sim              Não     Fibra optica               Não   

  servico_backup servico_protecao_equipamento servico_suporte_tecnico  \
0            Sim                          Não                     Não   
1            Não                          Sim                     Não   
2            Sim                          Não                     Não   
3            Não                          Sim                     Sim   
4            Não                          Não                     Não   

  streaming_tv streaming_filmes contrato fatura_digital forma_pagamento  \
0          Não              Não   Mensal            Sim          Boleto   
1          Não              Não    Anual            Não          Cheque   
2          Não              Não   Mensal            Sim          Cheque   
3          Não              Não    Anual            Não   Transferencia   
4          Não              Não   Mensal            Sim          Boleto   

   valor_mensal  total_gasto churn  
0         29.85        29.85   Não  
1         56.95      1889.50   Não  
2         53.85       108.15   Sim  
3         42.30      1840.75   Não  
4         70.70       151.65   Sim

Temos dois tipos de variáveis: numéricas e categóricas. As variáveis numéricas são aquelas que representam uma quantidade, como por exemplo, a quantidade de meses que o cliente permaneceu na empresa. Já as variáveis categóricas são aquelas que representam uma categoria, como por exemplo, o gênero do cliente. É útil separar as variáveis numéricas das variáveis categóricas para que possamos analisá-las separadamente.

In [8]:
colunas_numericas = df_churn.select_dtypes(include='number').columns
colunas_categoricas = df_churn.select_dtypes(exclude='number').columns

print(colunas_numericas)
print(colunas_categoricas) 

Index(['meses_na_empresa', 'valor_mensal', 'total_gasto'], dtype='object')
Index(['genero', 'idoso', 'tem_parceiro', 'tem_dependentes',
       'servico_telefone', 'multiplas_linhas', 'servico_internet',
       'servico_seguranca', 'servico_backup', 'servico_protecao_equipamento',
       'servico_suporte_tecnico', 'streaming_tv', 'streaming_filmes',
       'contrato', 'fatura_digital', 'forma_pagamento', 'churn'],
      dtype='object')


O `churn` é uma variável categórica, mas ela não será incluída na lista de variáveis categóricas porque ela é a variável alvo. Variável alvo é o nome dado à variável que queremos prever. Neste caso, queremos prever se um cliente vai cancelar o serviço ou não. Portanto, o `churn` não será incluído na lista de variáveis categóricas.

In [9]:
colunas_categoricas = colunas_categoricas.drop('churn')
print(colunas_categoricas)

Index(['genero', 'idoso', 'tem_parceiro', 'tem_dependentes',
       'servico_telefone', 'multiplas_linhas', 'servico_internet',
       'servico_seguranca', 'servico_backup', 'servico_protecao_equipamento',
       'servico_suporte_tecnico', 'streaming_tv', 'streaming_filmes',
       'contrato', 'fatura_digital', 'forma_pagamento'],
      dtype='object')
